rpc/src/v0_34/client/subscription.rs

```diff
@@ -0,0 +1,92 @@
+//! Subscription- and subscription management-related functionality.
+
+use core::pin::Pin;
+
+use async_trait::async_trait;
+use futures::{
+    task::{Context, Poll},
+    Stream,
+};
+use pin_project::pin_project;
+
+use crate::{client::subscription::Inner, prelude::*, query::Query, v0_34::event::Event, Error};
+
+/// A client that exclusively provides [`Event`] subscription capabilities,
+/// without any other RPC method support.
+#[async_trait]
+pub trait SubscriptionClient {
```

Comment:
One of the benefits of this approach here is that the API can evolve across different versions of Tendermint, but one of the drawbacks is that callers need to refer to Tendermint version-specific client traits.

An alternative approach here would be to have one `SubscriptionClient` trait for all supported versions of Tendermint, and have it be generic over the `Event` type produced by the subscription client. Then the caller could also be generic over the event type. I'd be interested to know whether you considered and ruled out such an approach? (I haven't actually done the hard work here, so I'm not aware yet of the ripple effects throughout the codebase of such an approach.)

Reply:
I think it's possible to introduce `Event` as an associated type for the `SubscriptionClient` trait, but I was not sure if the more substantial API change is desirable. Currently, `v0_34` can be treated like a bolt-on with no complications to the definitions elsewhere. In the relayer, we'll have to explicitly specialize for both event types anyway, because parsing for old subscription events is (unnecessarily) different.

Reply:
Makes sense :+1: I'd say we should keep your changes as-is then, and we can always look at merging the traits at some point in a future version of tendermint-rs if necessary.